File: src/main/java/org/kite9/diagram/common/objects/BasicBounds.java
package org.kite9.diagram.common.objects;

import java.text.DecimalFormat;
import java.text.NumberFormat;

import org.apache.commons.math.fraction.BigFraction;
import org.kite9.framework.logging.LogicException;

public class BasicBounds implements Bounds {
	
	public static final BasicBounds EMPTY_BOUNDS = new BasicBounds(-1, -1);

	
	private double min, max;

	@Override
	public Bounds expand(Bounds other) {
		if (this == EMPTY_BOUNDS) {
			return other;
		} else if (other == EMPTY_BOUNDS) {
			return this;
		}
		BasicBounds o2 = (BasicBounds) other;
		//System.out.println("merging "+o2+" with "+this);
		return new BasicBounds(Math.min(o2.getDistanceMin(), this.getDistanceMin()), 
				Math.max(o2.getDistanceMax(), this.getDistanceMax()));
	}
	
	@Override
	public Bounds narrow(Bounds other) {
		if (this == EMPTY_BOUNDS) {
			return this;
		} else if (other == EMPTY_BOUNDS) {
			return other;
		}
		BasicBounds o2 = (BasicBounds) other;
		double lower = Math.max(o2.getDistanceMin(), this.getDistanceMin());
		double upper = Math.min(o2.getDistanceMax(), this.getDistanceMax());

		if (lower >= upper) {
			return EMPTY_BOUNDS;
		}

		return new BasicBounds(lower, upper);
	}
	

	@Override
	public double getDistanceMin() {
		return min;
	}

	@Override
	public double getDistanceMax() {
		return max;
	}
	
	@Override
	public double getDistanceCenter() {
		return (getDistanceMax()+getDistanceMin())/2d;
	}

	public BasicBounds(double min, double max) {
		super();
		this.min = min;
		this.max = max;
		if (min > max) {
			throw new LogicException("Illegal Bounds");
		}
	}

	public static final NumberFormat nf = new DecimalFormat(".0000");

	public String toString() {
		return "(bb, g="+
				nf.format(getDistanceMin())+
				"-"+
				nf.format(getDistanceMax())+")";
	}

	/**
	 * As well as following the usual -1, 0, 1 compare operation, this also returns the common level of the two bounds 
	 * being compared.
	 */
	@Override
	public int compareTo(Bounds s) {
		if (getDistanceMax() < s.getDistanceMin()) {
			return -1;
		} else if (getDistanceMin() > s.getDistanceMax()) {
			return 1;
		} else {
			return 0;
		}
	}


	@Override
	public Bounds keep(double buffer, double width, BigFraction atFraction) {
		double span = max - min - (buffer * 2d);
		double pos = atFraction.doubleValue() * span;
		double lower = min + pos - (width / 2d) + buffer;
		double upper = min + pos + (width / 2d) + buffer;
		lower = Math.max(min+buffer, lower);
		upper = Math.min(max-buffer, upper);
		return new BasicBounds(lower, upper);
	}
	
	@Override
	public Bounds keep(double buffer, double width, double fraction) {
		double span = max - min - (buffer * 2d);
		double pos = fraction * span;
		double lower = min + pos - (width / 2d) + buffer;
		double upper = min + pos + (width / 2d) + buffer;
		lower = Math.max(min+buffer, lower);
		upper = Math.min(max-buffer, upper);
		return new BasicBounds(lower, upper);
	}

//	@Override
//	public Bounds keepMax(double lb, double ub) {
//		if ((lb == 0) && (ub == 0)) {
//			return this;
//		}
//		return new BasicBounds(this.max - ub, this.max - lb);
//	}
//
//
//	@Override
//	public Bounds keepMin(double lb, double ub) {
//		if ((lb == 0) && (ub == 0)) {
//			return this;
//		}
//		return new BasicBounds(this.min+lb ,this.min+ub);
//	}
//	
//
//	@Override
//	public Bounds keepMid(double w) {
//		double mid = (this.min + this.max) / 2d;
//		if ((mid <0) || (mid > 1)) {
//			return BasicBounds.EMPTY_BOUNDS;
//		}
//		return new BasicBounds(mid-(w/2),mid+(w/2));
//	}

}

File: src/main/kotlin/org/kite9/diagram/common/objects/BasicBounds.kt
package org.kite9.diagram.common.objects

import org.kite9.diagram.common.fraction.BigFraction
import org.kite9.diagram.logging.LogicException

data class BasicBounds(override val distanceMin: Double, override val distanceMax: Double) : Bounds {

    override fun expand(other: Bounds): Bounds {
        if (this === EMPTY_BOUNDS) {
            return other
        } else if (other === EMPTY_BOUNDS) {
            return this
        }
        val o2 = other as BasicBounds
        //System.out.println("merging "+o2+" with "+this);
        return BasicBounds(
            Math.min(o2.distanceMin, distanceMin),
            Math.max(o2.distanceMax, distanceMax)
        )
    }

    override fun narrow(other: Bounds): Bounds {
        if (this === EMPTY_BOUNDS) {
            return this
        } else if (other === EMPTY_BOUNDS) {
            return other
        }
        val o2 = other as BasicBounds
        val lower = Math.max(o2.distanceMin, distanceMin)
        val upper = Math.min(o2.distanceMax, distanceMax)
        return if (lower >= upper) {
            EMPTY_BOUNDS
        } else BasicBounds(lower, upper)
    }

    override val distanceCenter: Double
        get() = (distanceMax + distanceMin) / 2.0

    override fun toString(): String {
        return "(bb, g=%.2f - %.2f)".format(distanceMin, distanceMax);
    }

    /**
     * As well as following the usual -1, 0, 1 compare operation, this also returns the common level of the two bounds
     * being compared.
     */
    override operator fun compareTo(s: Bounds): Int {
        return if (distanceMax < s.distanceMin) {
            -1
        } else if (distanceMin > s.distanceMax) {
            1
        } else {
            0
        }
    }

    override fun keep(buffer: Double, width: Double, atFraction: BigFraction): Bounds {
        val span = distanceMax - distanceMin - buffer * 2.0
        val pos = atFraction.doubleValue() * span
        var lower = distanceMin + pos - width / 2.0 + buffer
        var upper = distanceMin + pos + width / 2.0 + buffer
        lower = Math.max(distanceMin + buffer, lower)
        upper = Math.min(distanceMax - buffer, upper)
        return BasicBounds(lower, upper)
    }

    override fun keep(buffer: Double, width: Double, fraction: Double): Bounds {
        val span = distanceMax - distanceMin - buffer * 2.0
        val pos = fraction * span
        var lower = distanceMin + pos - width / 2.0 + buffer
        var upper = distanceMin + pos + width / 2.0 + buffer
        lower = Math.max(distanceMin + buffer, lower)
        upper = Math.min(distanceMax - buffer, upper)
        return BasicBounds(lower, upper)
    }

    override fun narrow(vertexTrim: Double): Bounds {
        return BasicBounds(distanceMin + vertexTrim, distanceMax - vertexTrim)
    }

    companion object {
        @JvmField
		val EMPTY_BOUNDS = BasicBounds(-1.0, -1.0)
    }

    init {
        if (distanceMin > distanceMax) {
            throw LogicException("Illegal Bounds")
        }
    }
}